### PR TITLE
Refactor the parser somewhat

### PR DESCRIPTION
**Pull request by pennae**
# Motivation

the parser is currently not in great shape. it's tightly coupled to EvalState, it does more than just parsing, and it contains heaps of not entirely related code in the core `parser.y` file. let's refactor it a bit to make it easier to work with.

# Context

# Priorities and Process

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

### CODE REVIEW

File: src/libexpr/eval.cc
Comment (tomberek) [inline src/libexpr/eval.cc:2771]: I was not aware of this feature. (Ref: https://github.com/NixOS/nix/issues/7026)

File: src/libexpr/eval.cc
Comment (tomberek) [inline src/libexpr/eval.cc:24]: This oddity was already in the parser. Moving it here is directionally correct, further future work might refactor this more. It is for the NIX_PATH handling of "flake:something" URLs.

**Review by tomberek — COMMENTED** [review]
This looks to be exactly as advertised, refactoring the parser and separting concerns a bit better than before.

**Comment by thufschmitt** [comment]
Discussed during the Nix maintainers meeting on 2024-01-22.
Idea approved, assigned to @ericson2314 and @tomberek for further review

<details>
- Talked about @pennae's larger plans for a parser rewrite

- Exciting stuff: correctness, big performance improvements too.
  - Often 20% faster (just parsing, to be clear)
  - Not more memory usage

- @edolstra says in some ways this is similar to previous iterations of the parser implementation

- @roberth: We'll hold of changes to the parser if needed and sync with them and pennae. (also parser PRs are rare)

- @tomberek, @roberth: Let's make a sort of announcement issue.
  - Explain the intent
  - Make sure other Nix implementations (tvix, rnix) are informed.
  - We don't get too many parser PRs, but this could be hint for anyone thinking of making one to wait while such big changes are in flight.

- #9776 assigned to @tomberek & @ericson2314 for joint review.

- @pennae: notes more prep PRs are on the way; looking forward to seeing them!

- @tomberek: are we OK dependencies wise?
  - Loose 2 gain 1 sounds like a good trade!

- @roberth: How is compile time performance?
  - @pennae: Haven't tested, but feels fine
  - Splitting the very large file (from this prep PR) also helps a lot.
  - @roberth: sounds good. If it feels reasonable, that's good enough :).

- @ericson2314: Also excited for the test coverage improvements, independent of parser rewriting.

Goals/requirements
- Do not allow users to write new syntax that's currently invalid
   - at least not without any warning
- Unblock the new parser

Decision Options (decision is #2):
1. Merge 9722 first
    - easiest
2. Close 9722 (picked!!!)
    - next-easiest, but creates larger parser rules
3. re-write 9722 to disallow arbitrary numbers of "/"s
    - most work and changes behavior

Decision: the new PEGTL parser is "Idea Approved"

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:1]: For API docs
```suggestion
#pragma once
///@file
```

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:8]: Doxygen doc
```suggestion
/**
 * @note Storing a C-style `char *` and `size_t` allows us to avoid
 * having to define the special members that using string_view here
 * would implicitly delete.
 */
```

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:13]: 4-space indent
```suggestion
    const char * p;
    size_t l;
    bool hasIndentation;
    operator std::string_view() const { return {p, l}; }
```

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:9]:
```suggestion
struct StringToken
{
```

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:16]:
```suggestion
struct ParserLocation
{
```

File: src/libexpr/parser-state.hh
Comment (Ericson2314) [inline src/libexpr/parser-state.hh:39]:
```suggestion
struct ParserState
{
```

**Review by Ericson2314 — APPROVED** [review]
I tried confirming things were just moved around with `git diff --color-moved`. Looks good!

**Pull request merged**
